--- a/net/mac80211/wme.c
+++ b/net/mac80211/wme.c
@@ -215,9 +215,13 @@ u16 ieee80211_select_queue(struct ieee80
 
 	/* use the data classifier to determine what 802.1d tag the
 	 * data frame has */
-	qos_map = rcu_dereference(sdata->qos_map);
-	skb->priority = cfg80211_classify8021d(skb, qos_map ?
-					       &qos_map->qos_map : NULL);
+	/* MAC queues enabler patch by Gurjashan S. Pannu <gspannu@mail.uni-paderborn.de>, date Wed, 27 Jul 2016 14:20:08 +0200 */
+	// Commenting this out seems to disable, however, the IP ToS bit setting the packet priority
+
+	// Quick workaround to enable queueing in different AC (BE, BK, VO, VI)
+	// qos_map = rcu_dereference(sdata->qos_map);
+	// skb->priority = cfg80211_classify8021d(skb, qos_map ?
+	//				       &qos_map->qos_map : NULL);
 
  downgrade:
 	ret = ieee80211_downgrade_queue(sdata, sta, skb);
